### PR TITLE
feat: add first draft for CoAP MUD-URL payloads

### PR DESCRIPTION
**Pull request by pulsastrix**

### CODE REVIEW

File: draft-romann-mud-constrained.md [hunk 1]
```diff
@@ -214,20 +214,72 @@ Using additional Link-Format parameters and well-known URIs, this document
...
+    It also makes this resource discoverable for MUD receivers using the methods specified in [REF] <!-- TODO REF resource discovery -->. 
+2.  The MUD Receiver discovers the resource using the aforementioned methods.
+    Depending on the method of discovery, this could for example happen using a periodic scan for devices, e.g., by periodically requesting a well-known URI using multicast.
+    Other methods of discovery might also provide a mechanism to directly notify the Receiver of new devices, in which case this method SHOULD be preferred over periodic scanning.
+3.  The MUD Receiver retrieves the discovered resource for devices where the MUD controller does not have an up-to-date MUD-URL stored.
+    To do so, it performs a CoAP request for the discovered MUD-URL resource URI using the GET method, which is responded to with the appropriate payload.
+    Receivers MUST specify the desired payload formats using the Accept option <!-- TODO cite -->.
+    If multiple payload formats are specified, the Thing MUST prefer ones that provide a greater degree of authenticity protection (i.e., prefer CWTs over plaintext transmission).
```
Comment (JKRhb) [inline draft-romann-mud-constrained.md]: The Accept option is actually non-repeatable, so from my understanding, content negotiation would occur over multiple exchanges if the Thing cannot fulfill the initial request. Would something like this work instead?

```suggestion
    Receivers MUST specify their desired payload format using the Accept option {{Section 5.10.4 of RFC7252}}.
    During content negotiation, Receivers MUST start with requesting the Content-Format that provides the greatest degree of authenticity protection (i.e., prefer CWTs over plaintext transmission).
```

File: draft-romann-mud-constrained.md [hunk 2]
```diff
@@ -214,20 +214,72 @@ Using additional Link-Format parameters and well-known URIs, this document
...
-### Using the MUD-URL Resource (Thing-side)
-
-### Using the MUD-URL Submission Resource (Receiver-side)
+## MUD CoAP Payloads
+CoAP allows transmission of payloads in different formats.<!-- TODO this sentence seems clunky -->
+For the purposes of this specification, we will define two formats for transmitting MUD-URLs, which are suitable for different environments.
+MUD receivers that conform to this specification MUST support both formats.
+
+### Plain URL
+The easiest method of transmitting MUD-URLs is using a plain text payload containing only the MUD-URL.
```
Comment (JKRhb) [inline draft-romann-mud-constrained.md]:
```suggestion
The simplest method of transmitting MUD-URLs is using a plain text payload containing only the MUD-URL.
```

File: draft-romann-mud-constrained.md [hunk 3]
```diff
@@ -214,20 +214,72 @@ Using additional Link-Format parameters and well-known URIs, this document
...
+
+This flow can be divided into these general steps:
+1.  The MUD Receiver provides a CoAP resource that Things can submit their MUD-URLs to.
+    It also makes itself discoverable for Things using the methods specified in [REF] <!-- TODO Ref Discovery -->.
+2.  The Thing connects to the network. 
+    After connecting, it discovers the MUD-URL submission resource using the aforementioned methods.
+3.  The Thing submits the MUD-URL to the previously discovered URI.
+    To do so, it performs a CoAP request to the discovered URI with the POST method.
+    The MUD-URL is contained as the message payload in this request using one of the content formats defined in [REF] <!-- TODO MUD CoAP Payloads -->.
+    <!-- TODO allow receiver to limit acceptable content formats? -->
```
Comment (JKRhb) [inline draft-romann-mud-constrained.md]: Would something like this work?

```suggestion
    Receivers MAY limit their accepted Content-Formats to ones with a authentication level.
    In this case, a Receiver MUST respond with the 4.15 (Unsupported Content-Format) Response Code if it does not allow the use of a provided payload format for security reasons.
```

File: draft-romann-mud-constrained.md [hunk 4]
```diff
@@ -305,6 +335,48 @@ In the remainder of this section, we will outline potential use-cases and proced
 
 ## Receiver Behavior
 
+MUD receivers are assumed to be mostly non-constrained devices.
+Accordingly, this specification puts most of the implementation burden regarding support for flows and formats on the receivers, while keeping the requirements for Things as small as possible.
+In general, it is recommended that MUD receivers support as much of the specification as possible in order to support as many different Things as possible.
+
+### Discovery
+
+For the discovery process described in [](#general_discovery), the following considerations apply to MUD receivers:
```
Comment (JKRhb) [inline draft-romann-mud-constrained.md]:
```suggestion
For the discovery process described in {{#general_discovery}}, the following considerations apply to MUD receivers:
```

File: draft-romann-mud-constrained.md (hunk 3 above)
Comment (pulsastrix) [reply draft-romann-mud-constrained.md]:
```suggestion
    Receivers MAY limit their accepted Content-Formats to ones with an authentication level.
```

File: draft-romann-mud-constrained.md [hunk 5]
```diff
@@ -180,22 +180,110 @@ In the following, we will first outline these additional means for exposing MUD
...
+<!-- TODO required ciphers? -->
+### For the Thing-initiated Flow
+1.  After submitting the MUD-URL, the MUD Receiver parses the token.
+    If it detects Proof-of-Possession claims, the receiver MUST reply with a 4.01 (Unauthorized) CoAP response code and reject the token for now.
+    If the original submission request was not performed using CoAPS, the Receiver MUST also return a set of key information that can be used to establish a CoAP+DTLS session with it, as well as a CoAPS URI indicating the location of the secured submission resource.
+    <!-- TODO key format? -->
+2.  The Thing then repeats the submission request using its own key information, the Receiver's key information, and the provided URI.
+3.  During the DTLS handshake, the Receiver MUST require the Thing to use the Proof-of-Possession key included in the originally submitted token in order to successfully establish a DTLS session.
+4.  After the DTLS handshake, the Thing has proven possession of the key included in the CWT to the MUD Receiver, which can then treat the CWT as valid.
+<!-- TODO do we still want to re-send the CWT? -->
```
Comment (JKRhb) [inline draft-romann-mud-constrained.md]: Hmm, in the rendered view, this subsection is a bit broken formatting-wise.

image

Comment (pulsastrix) [reply draft-romann-mud-constrained.md]: fixed in [726cc52](https://github.com/namib-project/draft-coap-mud/pull/9/commits/726cc52c39983404b7c6ab25ca18aa16ab86cae2)

**Pull request merged**